src/helpers.c

```diff
@@ -149,130 +149,116 @@ void concat_str_array(char **output, char *const argv[])
 	}
 }
 
-int system_argv(char *const argv[])
+int system_argv_pipe(char *const lhscmd[], char *const rhscmd[])
 {
-	int child_exit_status;
-	pid_t pid;
-	int status = -1;
+	return system_argv_pipe_fd(-1, -1, lhscmd, -1, -1, rhscmd);
+}
 
-	pid = fork();
+int system_argv_pipe_fd(int lnewstdinfd, int lnewstderrfd, char *const lhscmd[],
+			int rnewstdoutfd, int rnewstderrfd, char *const rhscmd[])
+{
+	pid_t monitorpid = fork();
+	if (monitorpid == -1) {
+		LOG(NULL, "Failed to create child process to monitor pipe between", "command %s and command %s", lhscmd[0], rhscmd[0]);
+		return -1;
+	} else if (monitorpid == 0) {
+		pipe_monitor(lnewstdinfd, lnewstderrfd, lhscmd, rnewstdoutfd, rnewstderrfd, rhscmd);
+	}
+	return wait_process_terminate(monitorpid);
+}
 
-	if (pid == 0) { /* child */
-		execvp(*argv, argv);
-		LOG(NULL, "This line must not be reached", "");
-		assert(0);
-	} else if (pid < 0) {
-		LOG(NULL, "Failed to fork a child process", "");
+void pipe_monitor(int lnewstdinfd, int lnewstderrfd, char *const lhscmd[],
+		  int rnewstdoutfd, int rnewstderrfd, char *const rhscmd[])
+{
+	int pipefd[2];
+	if (pipe(pipefd) == -1) {
+		LOG(NULL, "Failed to create a pipe between", "command %s and command %s", lhscmd[0], rhscmd[0]);
```

Comment:
The comma between the 2 constant strings looks wrong. Without actually looking I would expect the second argument to LOG to be a printf style string.  If there is not a compile time warning we should update our headers to get it to do so.

Reply:
LOG is used is inconsistently in the code base.  I too think it makes more sense to model this after printf usage than with this alternative usage.  Here's the definition of LOG:

`#define LOG(file, msg, fmt...) __log_message(file, msg, __FILE__, __LINE__, fmt)`

I think we can safely remove the extra comma?

Reply:
Ah, OK, if that is the definition then your code was correct, and you should ignore me.

Reply:
OK, I'll go with the extra comma usage then.